compiler/rustc_passes/src/check_attr.rs

```diff
@@ -292,6 +293,56 @@ impl CheckAttrVisitor<'_> {
         }
     }
 
+    /// Checks if a `#[no_coverage]` is applied directly to a function
+    fn check_no_coverage(
+        &self,
+        hir_id: HirId,
+        attr: &Attribute,
+        span: Span,
+        target: Target,
+    ) -> bool {
+        match target {
+            // no_coverage on function is fine
+            Target::Fn
+            | Target::Closure
+            | Target::Method(MethodKind::Trait { body: true } | MethodKind::Inherent) => true,
+
+            // function prototypes can't be covered
+            Target::Method(MethodKind::Trait { body: false }) | Target::ForeignFn => {
+                self.tcx.struct_span_lint_hir(UNUSED_ATTRIBUTES, hir_id, attr.span, |lint| {
+                    lint.build("`#[no_coverage]` is ignored on function prototypes").emit();
+                });
+                true
+            }
+
+            Target::Mod | Target::ForeignMod | Target::Impl | Target::Trait => {
+                self.tcx.struct_span_lint_hir(UNUSED_ATTRIBUTES, hir_id, attr.span, |lint| {
+                    lint.build("`#[no_coverage]` cannot be done recursively and must be applied to functions directly").emit();
```

Comment:
Nit: it isn't super clear to me what the intent is behind the recursion reference. I suspect you meant to point out that the attribute does not _propagate_ so perhaps something like:

```suggestion
                    lint.build("`#[no_coverage]` attribute does not propagate and must be applied directly to the contained functions").emit();
```

Reply:
I went with:

> `#[no_coverage]` does not propagate into items and must be applied to the contained functions directly

Does that seem good?